returnn/datasets/basic.py

```diff
@@ -967,6 +967,25 @@ def num_seqs(self) -> int:
         """
         raise NotImplementedError
 
+    def get_epoch_continuous(self, sorted_seq_idx: int) -> float:
+        """
+        Calculates the position of ``sorted_seq_idx`` in the current epoch as relative value in [0, 1].
+        ``sorted_seq_idx`` cannot be less than the seq index of the previously loaded seq.
+
+        This value is used to calculate ``epoch_continuous`` for any dynamic learning rate scheduling.
+
+        :param sorted_seq_idx: sorted seq idx
+        :return: continuous value in [0, 1] which represents the position in the current epoch
```

Comment:
```suggestion
        :return: continuous value in [0, 1) which represents the position in the current epoch
```
It's actually important that this never reaches 1, so that it can easily be used to calculate `epoch_continuous` based on this, and such that we can guarantee that `int(epoch_continuous) == epoch`, which would otherwise not be the case for the last seq of the dataset.

Reply:
We actually do this differently in the torch engine right now. There the value goes from `[epoch - 1, epoch]` inclusive.

Reply:
In the torch engine we have (as you correctly say):
```
epoch_continuous = self.epoch - 1 + (last_seq_idx + 1) / num_seqs
```
Why does this mean this value can never reach 1? It must, eventually, if `seq_idx == num_seqs - 1`.

Reply:
Ah yes sorry, I confused this. So it's actually (0,1]. The interpretation is: How much have we finished of the epoch including the seqs of the current batch. We start with the first seq, which is close to 0.0, and the last minibatch of the first epoch (epoch 1) would then finally reach 1.0.

We should not change how it is in the engine now. So what we do here should be consistent to that.

Reply:
It is consistent to the existing behavior the way it is now!